library/core/src/sync/atomic.rs

```diff
@@ -3326,23 +3326,23 @@ macro_rules! atomic_int {
                 unsafe { atomic_xor(self.v.get(), val, order) }
             }
 
-            /// Fetches the value, and applies a function to it that returns an optional
-            /// new value. Returns a `Result` of `Ok(previous_value)` if the function returned `Some(_)`, else
-            /// `Err(previous_value)`.
+            /// Loads the current value, applies a closure to it, and optionally tries to store a new value.
             ///
-            /// Note: This may call the function multiple times if the value has been changed from other threads in
-            /// the meantime, as long as the function returns `Some(_)`, but the function will have been applied
-            /// only once to the stored value.
+            /// If the closure ever returns None, this method will immediately return `Err(current value)`.
+            /// If the closure returns Some(new value), then this method calls
+            #[doc = concat!("[`", stringify!($atomic_type), "::compare_exchange_weak`]")]
+            /// to try to store the new value.
             ///
-            /// `fetch_update` takes two [`Ordering`] arguments to describe the memory ordering of this operation.
-            /// The first describes the required ordering for when the operation finally succeeds while the second
-            /// describes the required ordering for loads. These correspond to the success and failure orderings of
-            #[doc = concat!("[`", stringify!($atomic_type), "::compare_exchange`]")]
-            /// respectively.
+            /// If storing a new value fails (due to another thread changing the current value),
+            /// then the given closure will be called again on the new current value
+            /// (returned by `compare_exchange_weak`), until either the closure returns None,
+            /// or `compare_exchange_weak` succeeds in storing a new value.
```

Comment:
I think we should try to talk about the semantics a bit more rather than directly to the implementation, something like

> If the value of the atomic observed when attempting to store the `updated_value` is not equal to the value from the initial load, then the store will not happen. Instead, the closure will be invoked again, this time with the new value observed during the store attempt. This pattern will continue in a loop until either the closure returns `None` or the store succeeds.

Reply:
Hmm, that seems very attractive. I will spend some time exploring this.